contracts/Groth16Verifier.sol

```diff
@@ -20,107 +20,66 @@
 pragma solidity 0.8.23;
 import "./Groth16.sol";
 
-library Pairing {
-  // The prime q in the base field F_q for G1
-  uint private constant _Q =
+contract Groth16Verifier is IGroth16Verifier {
+  uint private constant _P =
     21888242871839275222246405745257275088696311157297823662689037894645226208583;
+  uint256 private constant _R =
+    21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+  VerifyingKey private _verifyingKey;
 
-  /// The negation of p, i.e. p.addition(p.negate()) should be zero.
-  function negate(G1Point memory p) internal pure returns (G1Point memory) {
-    if (p.x == 0 && p.y == 0) return G1Point(0, 0);
-    return G1Point(p.x, _Q - (p.y % _Q));
+  struct VerifyingKey {
+    G1Point alpha1;
+    G2Point beta2;
+    G2Point gamma2;
+    G2Point delta2;
+    G1Point[] ic;
   }
 
-  /// The sum of two points of G1
-  function addition(
-    G1Point memory p1,
-    G1Point memory p2
-  ) internal view returns (G1Point memory r) {
-    uint[4] memory input;
-    input[0] = p1.x;
-    input[1] = p1.y;
-    input[2] = p2.x;
-    input[3] = p2.y;
-    bool success;
-    // solhint-disable-next-line no-inline-assembly
-    assembly {
-      success := staticcall(sub(gas(), 2000), 6, input, 0xc0, r, 0x60)
```

Comment:
These `sub(gas(),2000)` things look very strange. I guess that would mean to pass the "available amount minus 2000" gas to the precompile? I guess technically it should work, but why 2000? or why subtract anything at all?

According to the docs I found, the first parameter of `staticcall` is

>  gas: amount of gas to send to the sub context to execute. The gas that is not used by the sub context is returned to this one.

Based on this just `gas()` would work? Also, what even does `gas()` mean, I cannot find it in the documentation... Is it maybe a deprecated name for `gasleft()`?

Reply:
Btw, the pairing precompile documentation says that 

"If the input is not valid, all gas provided is consumed and there is no return data, indicating a precompile contract error."

so maybe you shouldn't actually pass all the gas available, because that makes failures more expensive?
(not that there should be failures in normal working, why would anybody spend gas by sending an incorrect proof...)

Reply:
Good question, I left this unchanged from the original code, but I don't know why they subtract 2000 either.

The `gas()` call is this EVM opcode: https://www.evm.codes/#5a?fork=shanghai
It should indeed return gas left.

Reply:
Oh I was looking at the wrong place, this is assembly opcode and I was looking at the solidity builtins...

Still the minus 2000 doesn't make much sense to me, but I'm very much a non-expert in EVM :)

Reply:
I looked into this a bit further, and decided to just go with `gas()` as you suggested.

Even though EIP-196 states that it "consumes all gas provided" when the inputs are invalid, I cannot reproduce that on hardhat, and looking through the Geth code, it also doesn't seem to do that (https://github.com/ethereum/go-ethereum/blob/b80643b7370075262fd6dfad7ae8aa77710e2ef1/core/vm/contracts.go#L484). 
Also, strangely, the accompanying EIP-197 does not state that it consumes all gas, even though it is a similar use case.

And in the end it doesn't matter whether or not it consumes all the gas. If it does, then verification fails with an 'out-of-gas' error, and if it doesn't, then `verify` returns false. In both cases the proof is not verified.